kotlinx-coroutines-core/common/src/flow/operators/Share.kt

```diff
@@ -139,14 +181,31 @@ private fun <T> CoroutineScope.launchSharingDeferred(upstream: Flow<T>, result:
  */
 @ExperimentalCoroutinesApi
 public fun <T> MutableSharedFlow<T>.asSharedFlow(): SharedFlow<T> =
-    object : SharedFlow<T> by this {}
+    ReadonlySharedFlow(this)
 
 /**
  * Represents this mutable state flow as read-only state flow.
  */
 @ExperimentalCoroutinesApi
 public fun <T> MutableStateFlow<T>.asStateFlow(): StateFlow<T> =
-    object : StateFlow<T> by this {}
+    ReadonlyStateFlow(this)
+
+private class ReadonlySharedFlow<T>(
+    flow: SharedFlow<T>
+) : SharedFlow<T> by flow, CancellableFlow<T>, FusibleFlow<T> {
+    override fun fuse(context: CoroutineContext, capacity: Int, onBufferOverflow: BufferOverflow) =
+        fuseSharedFlow(context, capacity, onBufferOverflow)
+}
+
+private class ReadonlyStateFlow<T>(
+    flow: StateFlow<T>
+) : StateFlow<T> by flow, CancellableFlow<T>, FusibleFlow<T>, DistinctFlow<T> {
+    override val isDefaultEquivalence: Boolean
+        get() = true
+
+    override fun fuse(context: CoroutineContext, capacity: Int, onBufferOverflow: BufferOverflow) =
+        fuseStateFlow(context, capacity, onBufferOverflow)
+}
```

Comment:
That's very nice to protect internals from casting to the mutable version, or fusing to get back to an unprotected mutable version.

When this gets released, it can be nice to have an IDE inspection that spots upcasts to the non mutable version, and suggest using `asXxxxFlow` extension for extra safety.

Reply:
👍 https://youtrack.jetbrains.com/issue/KT-39443